fud2/fud-core/src/script/plugin.rs

```diff
@@ -1,150 +1,191 @@
 use crate::{
-    config,
     exec::{OpRef, SetupRef, StateRef},
     DriverBuilder,
 };
-use std::{cell::RefCell, path::PathBuf, rc::Rc};
+use std::{
+    cell::RefCell,
+    path::{Path, PathBuf},
+    rc::Rc,
+};
 
 use super::{
     error::RhaiSystemError,
     exec_scripts::{to_rhai_err, to_str_slice, RhaiResult, RhaiSetupCtx},
     report::RhaiReport,
 };
 
-fn to_setup_refs(
-    ctx: &rhai::NativeCallContext,
-    setups: rhai::Array,
+struct ScriptContext {
+    builder: DriverBuilder,
     path: PathBuf,
     ast: rhai::AST,
-    this: Rc<RefCell<DriverBuilder>>,
-) -> RhaiResult<Vec<SetupRef>> {
-    setups
-        .into_iter()
-        .map(|s| match s.clone().try_cast::<rhai::FnPtr>() {
-            Some(fnptr) => Ok(this.borrow_mut().add_setup(
-                &format!("{} (plugin)", fnptr.fn_name()),
-                RhaiSetupCtx {
-                    path: path.clone(),
-                    ast: ast.clone(),
-                    name: fnptr.fn_name().to_string(),
-                },
-            )),
-            // if we can't cast as a FnPtr, try casting as a SetupRef directly
-            None => s.clone().try_cast::<SetupRef>().ok_or_else(move || {
-                RhaiSystemError::setup_ref(s)
-                    .with_pos(ctx.position())
-                    .into()
-            }),
-        })
-        .collect::<RhaiResult<Vec<_>>>()
 }
 
-pub trait LoadPlugins {
-    fn load_plugins(self) -> Self;
+impl ScriptContext {
+    /// Take a Rhai array value that is supposed to contain setups and produce
+    /// an array of actual references to setups. The array might contain string names
+    /// for the setups, or it might be function references that define those setups.
+    fn setups_array(
+        &mut self,
+        ctx: &rhai::NativeCallContext,
+        setups: rhai::Array,
+    ) -> RhaiResult<Vec<SetupRef>> {
+        setups
+            .into_iter()
+            .map(|s| match s.clone().try_cast::<rhai::FnPtr>() {
+                Some(fnptr) => {
+                    // TODO: Do we really need to clone stuff here, or can we continue to thread through
+                    // the `Rc`?
+                    let rctx = RhaiSetupCtx {
+                        path: self.path.clone(),
+                        ast: self.ast.clone(),
+                        name: fnptr.fn_name().to_string(),
+                    };
+                    Ok(self.builder.add_setup(
+                        &format!("{} (plugin)", fnptr.fn_name()),
+                        rctx,
+                    ))
+                }
+                // if we can't cast as a FnPtr, try casting as a SetupRef directly
```

Comment:
if we are going to be stringly typed anyways with `get_setup`, should we consider also allowing passing strings directly here?

Reply:
Yeah, good point. Maybe that should work for states as well… as in, in a call to `op`, an actual `StateRef` works just as well as a string name for the op.